### PR TITLE
Standardize semantic segmentation models outputs

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR standardizes the model outputs for semantic segmentation models and creates an `AutoModelForSemanticSegmentation` class. As discussed internally, models for semantic segmentation should return `logits` of shape `batch_size, num_labels, height, width`, one logit per pixel.

**Breaking change:** The `BeitForSemanticSegmentation` and `SegformerForSemanticSegmentation` models have logits with the same height and width as the input after this PR (instead of height/4 and width/4).

To maintain some level of backward compatibility, the `SemanticSegmentationModelOutput` has a field `legacy_logits` that users can pick to get the old logits value.

Another possible road to be less breaking is to create new classes `BeitForPixelClassification` and `SegformerForPixelClassification` while deprecating the current ones, then name "instance segmentation" "pixel classification" everywhere. It has the benefit of being more understandable to the beginner and look like our "ForTokenClassification" classes but it might surprise an expert more used to the "semantic segmentation" name.

### CODE REVIEW

File: src/transformers/modeling_outputs.py [hunk 1]
```diff
@@ -812,3 +812,36 @@ class Seq2SeqQuestionAnsweringModelOutput(ModelOutput):
     encoder_last_hidden_state: Optional[torch.FloatTensor] = None
     encoder_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
     encoder_attentions: Optional[Tuple[torch.FloatTensor]] = None
+
+
+@dataclass
+class SemanticSegmentationModelOutput(ModelOutput):
+    """
+    Base class for outputs of instance segmentation models.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:820]:
```suggestion
    Base class for outputs of semantic segmentation models.
```

File: src/transformers/modeling_outputs.py [hunk 2]
```diff
@@ -812,3 +812,36 @@ class Seq2SeqQuestionAnsweringModelOutput(ModelOutput):
...
+
+    Args:
+        loss (`torch.FloatTensor` of shape `(1,)`, *optional*, returned when `labels` is provided):
+            Classification (or regression if config.num_labels==1) loss.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:824]: Do we also want to tackle regression tasks with this output class?

Depth estimation is an example of per-pixel regression, but that's a different task.

**Review by NielsRogge — COMMENTED** [review]
Looks good to me (wondering why the PR is called "standardize instance segmentation models outputs"?).

* `xxxForPixelClassification` => both semantic and panoptic segmentation are pixel classification tasks, right? Although in semantic segmentation, one only predicts a single label per pixel, whereas one predicts two labels per pixel in case of panoptic segmentation. If we're going with `xxxForPixelClassification` and `xxxForPanopticSegmentation`, people would be confused and that's not really consistent. So I would stick to `xxxForSemanticSegmentation`.

There's also depth estimation, which is per-pixel regression (if I understand correctly, I'm not entirely familiar with it). Would we then have `xxxForPixelRegression`? However, currently we also tackle both regression and classification with `xxxForSequenceClassification` models.

File: src/transformers/modeling_outputs.py [hunk 3]
```diff
@@ -812,3 +812,36 @@ class Seq2SeqQuestionAnsweringModelOutput(ModelOutput):
...
+        logits (`torch.FloatTensor` of shape `(batch_size, config.num_labels, height, width)`):
+            Classification scores for each pixel.
+        legacy_logits (`torch.FloatTensor` of shape `(batch_size, config.num_labels, height/4, width/4)`):
+            For backward compatibility, old logits returned by `BeitForSemanticSegmentation` and
+            `SegformerForSemanticSegmentation`.
```
Comment (NielsRogge) [inline src/transformers/modeling_outputs.py:829]: Not sure if we should keep this. We could also call it `raw_logits` (which are the raw logits coming out of the model without any interpolation).

Comment (FrancescoSaverioZuppichini) [reply src/transformers/modeling_outputs.py:829]: I have a doubt about the naming. `logits` is usually used to indicate un-normalized vectors used to compute a downstream task (e.g. classification after a softmax layer). If that is the case, return this output class is unfeasible for some models, e.g. `MaskFormer`. 

In `MaskFormer` we compute the final segmentation masks using two vectors

```python
def forward(self, *args, **kwargs):
        outputs: MaskFormerOutput = self.model(*args, **kwargs)
        # mask classes has shape [BATCH, QUERIES, CLASSES + 1]
        # remove the null class `[..., :-1]`
        masks_classes: Tensor = outputs.preds_logits.softmax(dim=-1)[..., :-1]
        # mask probs has shape [BATCH, QUERIES, HEIGHT, WIDTH]
        masks_probs: Tensor = outputs.preds_masks.sigmoid()
        # now we want to sum over the queries,
        # $ out_{c,h,w} =  \sum_q p_{q,c} * m_{q,h,w} $
        # where $ softmax(p) \in R^{q, c} $ is the mask classes
        # and $ sigmoid(m) \in R^{q, h, w}$ is the mask probabilities
        # b(atch)q(uery)c(lasses), b(atch)q(uery)h(eight)w(idth)
        segmentation: Tensor = torch.einsum("bqc, bqhw -> bchw", masks_classes, masks_probs)

        return MaskFormerForSemanticSegmentationOutput(segmentation=segmentation, **outputs)
```        
Returning a vector `logits` of shape `batch_size, config.num_labels, height, width)` is impossible due. I see two possible solutions

**Review by FrancescoSaverioZuppichini — COMMENTED** [review]
Thanks! I've added a comment about `logits` in the output and why it may be challenging for `MaskFormer` to use it

File: src/transformers/modeling_outputs.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/modeling_outputs.py:824]: We can add it to the docstring once it's supported. For now, the models don't use `problem_type` to infer the proper loss function.

File: src/transformers/modeling_outputs.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_outputs.py:829]: We might have to treat `MaskFormer` with an exception (in the sense we will need the raw output to have both logits and prediction masks to be able to compute the segmentation) the same way we handle the model for QA with beam search in XLNert. That class would then not be in the `AutoModelForSemanticSegmentation` as it doesn't have the same API, but we would need one version of `MaskFormerForSemanticSegmentation` with the same API (it might perform as well as this fancy one with some fine-tuning).

File: src/transformers/models/auto/modeling_auto.py [hunk 4]
```diff
@@ -278,11 +278,20 @@
...
 )
 
+MODEL_FOR_SEMANTIC_SEGMENTATION_MAPPING_NAMES = OrderedDict(
+    [
+        # Model for Semantic Segmentation mapping
+        ("beit", "BeitForSemanticSegmentation"),
+        ("segformer", "SegformerForSemanticSegmentation"),
+    ]
+)
+
```
Comment (LysandreJik) [inline src/transformers/models/auto/modeling_auto.py:294]: LGTM

**Review by LysandreJik — COMMENTED** [review]
I don't have sufficient experience with computer vision and these specific tasks to review the content of the PR, so my review is on the:

- coherence of this auto model with the rest of the auto models 
- backwards compatibility

I think this PR is great as it's a good opportunity to observe if our "experimental" approach works or not. I think it shows that it lacks quite a few aspects that would make it robust:

- We do not systematically mention that models are experimental on their modeling page, which is an issue. It should be very visible from the users.
- This is an urgent refactor, but ideally this should go through a correct deprecation cycle. I think we need to be clear that this is an exceptional instance where we do not respect backwards compatibility for experimental reasons.
- In order to patch this, I think we'll need to:
    - Tweet (can be from personal accounts) about the update
    - Create a forum post
    - Create a pinned github issue

Thanks for working on this, @sgugger!

File: src/transformers/models/beit/modeling_beit.py [hunk 5]
```diff
@@ -1180,13 +1190,21 @@ def forward(
...
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
         output_hidden_states = (
             output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
         )
+        legacy_output = legacy_output if legacy_output is not None else self.config.legacy_output
+        if not legacy_output:
+            warnings.warn(
+                "The output of this model has changed in v4.17.0 and the logits now have the same size as the inputs. "
+                "You can activate the previous behavior by passing `legacy_output=True` to this call or the "
+                "configuration of this model (only until v5, then that argument will be removed.",
```
Comment (LysandreJik) [inline src/transformers/models/beit/modeling_beit.py:1205]:
```suggestion
                "configuration of this model (only until v5, then that argument will be removed).",
```

**Review by LysandreJik — APPROVED** [review]
Thank you for iterating, now it looks good to me from a backwards compatibility perspective.
